contracts/sponsorship/BiconomySponsorshipPaymaster.sol

```diff
@@ -261,25 +265,29 @@ contract BiconomySponsorshipPaymaster is
         override
     {
         unchecked {
-            (address paymasterId, uint32 priceMarkup, bytes32 userOpHash, uint256 prechargedAmount) =
-                abi.decode(context, (address, uint32, bytes32, uint256));
+            (address paymasterId, uint32 priceMarkup, uint256 prechargedAmount) =
+                abi.decode(context, (address, uint32, uint256));
 
             // Include unaccountedGas since EP doesn't include this in actualGasCost
             // unaccountedGas = postOpGas + EP overhead gas + estimated penalty
             actualGasCost = actualGasCost + (unaccountedGas * actualUserOpFeePerGas);
             // Apply the price markup
             uint256 adjustedGasCost = (actualGasCost * priceMarkup) / _PRICE_DENOMINATOR;
 
+            uint256 premium = adjustedGasCost - actualGasCost;
+
+            // Add priceMarkup to fee collector balance
+            paymasterIdBalances[feeCollector] += premium;
+
             if (prechargedAmount > adjustedGasCost) {
                 // If overcharged refund the excess
                 paymasterIdBalances[paymasterId] += (prechargedAmount - adjustedGasCost);
+                // here adjustedGasCost does not account for gasPenalty
+                emit GasBalanceDeducted(paymasterId, adjustedGasCost, premium);
+            } else {
+                // here chargedAmount accounts for penalty with maxGasPenalty
+                emit GasBalanceDeducted(paymasterId, prechargedAmount, premium);
```

Comment:
can it really happen? 
if `prechargedAmount < adjustedGasCost` doesn't paymaster lose money?

if  we only expect `else` to mean` prechargedAmount == adjustedGasCost`, we do need `else` at all

Reply:
I feel it would never go into the else case, but if it does then want to emit different value in the event.

Reply:
if we pre charged less than we want to charge more yeah. in that case otherwise paymaster loses money. so diff of that adjusted - pre charged should be deduced from paymasterId balance.

Reply:
Makes sense. Does it make sense to emit different event to easily recognize we charged less

Reply:
Can do it but it will be overhead on the paymaster service. 
so, should I leave this 
1. as it is
2. as it is but different event name
3. same/different event name but in the else case doing
    paymasterIdBalances[paymasterId] += (adjustedGasCost - prechargedAmount);

Reply:
@filmakarov

Reply:
going with 3 with same event name.